ap/inference/datacard.py

```diff
@@ -0,0 +1,550 @@
+                            # skip symmetric effects
+                            if not isinstance(effect, tuple) and len(effect) != 2:
+                                continue
+                            # skip one sided effects
+                            if not (min(effect) <= 1 <= max(effect)):
+                                continue
+                            d, u = effect
+                            effect = 0.5 * (u - d) + 1.0
+
+                        elif trafo == ParameterTransformation.asymmetrize or (
+                            trafo == ParameterTransformation.asymmetrize_if_large and
+                            isinstance(effect, float) and
+                            abs(effect - 1.0) >= 0.2
+                        ):
+                            # skip asymmetric effects
+                            if not isinstance(effect, float):
+                                continue
+                            effect = (2.0 - effect, effect)
+
+                elif _param_obj.type.is_shape:
+                    # when the shape was constructed from a rate, reset the effect to 1
+                    if _param_obj.transformations.any_from_rate:
+                        effect = 1.0
+
+                # encode the effect
+                if isinstance(effect, (int, float)):
+                    if effect == 0.0:
+                        effects.append("-")
+                    elif effect == 1.0 and _param_obj.type.is_shape:
+                        effects.append("1")
+                    else:
+                        effects.append(str(rnd(effect)))
+                elif isinstance(effect, tuple) and len(effect) == 2:
+                    effects.append(f"{rnd(effect[0])}/{rnd(effect[1])}")
+                else:
+                    raise ValueError(
+                        f"effect '{effect}' of parameter '{param_name}' with type {param_obj.type} "
+                        f"on process '{proc_name}' in category '{cat_name}' cannot be encoded",
+                    )
+
+            # add the tabular line
+            if param_obj and effects:
+                type_str = "shape"
+                if param_obj.type == ParameterType.rate_gauss:
+                    type_str = "lnN"
+                elif param_obj.type == ParameterType.rate_uniform:
+                    type_str = "lnU"
+                blocks.tabular_parameters.append([param_name, type_str, effects])
+
+        if blocks.tabular_parameters:
+            empty_lines.add("tabular_parameters")
+
+        # line-style parameters
+        blocks.line_parameters = []
+        for param_name in self.inference_model_inst.get_parameters(flat=True):
+            for cat_name, proc_name in flat_rates:
+                param_obj = self.inference_model_inst.get_parameter(
+                    param_name,
+                    category=cat_name,
+                    process=proc_name,
+                    silent=True,
+                )
```

Comment:
Ah, I see :+1: Though you don't really have to loop through _all_ parameters, right? You already skip the rateParams (as you should of course), so I was thinking of extending the getter functions to allow a keyword `type` or something like that that also allows patterns/regex. For example, I could imagine that it can be usecases where you just want to do something with shape variations, or all constrained parameters (i.e. (shape|lnN) or something). It would be practical to then only get the parameters you're actually interested in imho. Of course, this is also no show stopper, and if we want to implement this at a later time (if at all) I'm also fine with that.

Reply:
Yeah, maybe better at a later stage. The types are enum flags at the moment, and I would have to think about a way to filter them using keyword args